### PR TITLE
stylo: Implement -moz-prefixed gradients

### PR DESCRIPTION
**Pull request by canova**
---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix [Bug 1366544](https://bugzilla.mozilla.org/show_bug.cgi?id=1366544)

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/17434)

### CODE REVIEW

File: components/style/values/computed/image.rs [hunk 1]
```diff
@@ -54,6 +54,8 @@ pub enum LineDirection {
     Angle(Angle),
     /// A corner.
     Corner(X, Y),
+    /// A Position and an Angle for legacy `-moz-` prefixed gradient.
+    PositionAngle(Option<Position>, Option<Angle>),
```
Comment (nox) [inline components/style/values/computed/image.rs]: Call that `MozPosition` and put `#[cfg(feature = "gecko")]` on it and on every code that uses it.

Comment (nox) [inline components/style/values/computed/image.rs]: Why are both parts optional?

File: components/style/values/specified/position.rs [hunk 2]
```diff
@@ -49,25 +49,46 @@ define_css_keyword_enum! { Y:
...
+/// But for some historical reasons we need to keep CSS Level 2 syntax which only supports up to
+/// 2-value. This enum represents whether position should parse modern syntax or legacy syntax.
+#[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum PositionSyntax {
+    /// Modern syntax
+    Modern,
+    /// Legacy syntax
+    Legacy,
+}
```
Comment (nox) [inline components/style/values/specified/position.rs]: What? Why is this needed? Please make a specific type if you need changes to `Position`,  this type is already complex as it is and I don't want any more hacks in it.

**Review by nox — CHANGES_REQUESTED** [review]

File: components/style/values/computed/image.rs (hunk 1 above)
Comment (canaltinova) [inline components/style/values/computed/image.rs]: Currently `-webkit-` prefixed gradients are also a servo gradient type. And I assumed we would want these in servo too. Am I wrong?

Comment (canaltinova) [inline components/style/values/computed/image.rs]: Because it can be position, angle or both. Position can be omitted when angle is specified or angle can be omitted when position is specified. But we check and return Errr in parsing code if both are None.

File: components/style/values/specified/position.rs (hunk 2 above)
Comment (canaltinova) [inline components/style/values/specified/position.rs]: The both positions are very similar. (In fact our position was like the legacy one once IIRC). Legacy one accepts 1 value components and modern one accepts 1 and 2 value components. I think moving this to another type and writing the nearly same but reduced parsing/serialization code would be a big code duplication.

File: components/style/values/computed/image.rs (hunk 1 above)
Comment (emilio) [inline components/style/values/computed/image.rs]: Yes, I don't think we should parse `-moz` gradients without a clear reason to. We already fail to parse other stuff like `@-moz-keyframes`.

Comment (canaltinova) [inline components/style/values/computed/image.rs]: Ok, updated the code.

**Comment by canaltinova** [comment]
I added also moz prefixed radial gradients. Can you look at that too? I added cfg's to parsing branches but I don't think adding cfg's to generic Gradient and GradientKind will do any good. If we do that we will need to also add their ToCss and other impls separately.

File: components/style/values/specified/position.rs (hunk 2 above)
Comment (nox) [inline components/style/values/specified/position.rs]: This is not a big code duplication, this is making sure we can't have values that aren't even representable on the CSS front. Cf. for example, `style::values::specified::transform::OriginComponent`.

**Comment by canaltinova** [comment]
@nox I changed because you asked but I have to say that I'm not very happy with the outcome of the new type. There is massive code duplication and workarounds with cfg's.

**Comment by nox** [comment]
Isn't `LegacyPositionComponent` just the same as `OriginComponent`?

**Comment by canaltinova** [comment]
Ooh, you're right. I haven't looked at this type before. This reduced the duplication indeed. Updated the code.

**Comment by canaltinova** [comment]
It seems like new codes have been added to codebase. Rebased.
@bors-servo r=Manishearth,nox
